### DISCUSSION TITLE
Cholesky decomposition

### DISCUSSION THREAD

**Discussion by EnricaBelfiore** [luca-fiorito-11/sandy#133] [General]
@luca-fiorito-11 @AitorBengoechea 

I and @GrimFe noticed that the last modification introduced in the sampling procedure does not work for all nuclear data covariance matrices. More in the details, we tested it for covariances of Pu239, U235, U233 and H1.
First of all, was this method already tested with real test cases? In case it was, which nuclides did you use for testing @AitorBengoechea?

We think that this problem is related to the fact that the covarinace for nuclear data are not always positive defined. As far as we understand, `to_positive` method handles this by transforming such matrices into SEMI-positive defined matrices. LU decomposition implemented inside `sparse_table_cholesky` method does not properly work for singular matrix (det=0).
 
We propose to back up to the previous decomposition scheme (**cov = U*D*U.T**) and to return **L=U*sqrt(D)**, so that **cov=L*L.T**, where cov is the semipositive approximation of the nuclear data covariance matrix. 
Unfortunately, `scipy.linalg.ldl` (implementing UDU.T decomposition) does not handle sparse matrix. Therefore our suggestion is to directly implement `scipy.linalg.ldl` decomposition in `get_L`. Further possible memory optimization can and should be considered.

**Comment by GrimFe** [comment]
Correct implementation of the covariance matrix decomposition is also needed to test the convergence of mean/std for the sampling according to a lognormal distribution.

**Comment by luca-fiorito-11** [comment]
@EnricaBelfiore ,this is very helpful! Can you report here a simple test where the current method fails.
@AitorBengoechea , can you implement Enrica's test in a PR and if needed go back to the previous QR decomposition?

**Reply by EnricaBelfiore** [reply]
This is a simple test that we did with the covariance of the hydrogen:
 
`endf6 = sandy.get_endf6_file('jeff_33','xs', 10010)`
`Cov = sandy.XsCov.from_endf6(endf6)`
`Cov = sandy.CategoryCov(Cov)`
`Cov.sampling(20)`
_RuntimeError: Factor is exactly singular_

**Comment by luca-fiorito-11** [comment]
I see three methods already developed.

> non-sparse [QR](https://github.com/luca-fiorito-11/sandy/commit/5bbb4c96703aaa006050b29f43c504853a7441de#diff-931b78d42ea4d37612c872368233d51ab33a22923b40fb856c21cf1105a7d11dL602) 
> non-sparse [LU](https://github.com/luca-fiorito-11/sandy/commit/5bbb4c96703aaa006050b29f43c504853a7441de#diff-931b78d42ea4d37612c872368233d51ab33a22923b40fb856c21cf1105a7d11dL1552)
> sparse [LU](https://github.com/luca-fiorito-11/sandy/blob/b0fdf73bc8a37f1fefa1379e4cf86cabc6235d2d/sandy/core/cov.py#L2178) 

Is sparsity needed?

How do we handle negative eigenvalues?
Aitor developed [to_positive](https://github.com/luca-fiorito-11/sandy/blob/b0fdf73bc8a37f1fefa1379e4cf86cabc6235d2d/sandy/core/cov.py#L1545)  to make them zero.
Other two functions [_reduce_size](https://github.com/luca-fiorito-11/sandy/blob/b0fdf73bc8a37f1fefa1379e4cf86cabc6235d2d/sandy/core/cov.py#L389)  and [_restore_size](https://github.com/luca-fiorito-11/sandy/blob/b0fdf73bc8a37f1fefa1379e4cf86cabc6235d2d/sandy/core/cov.py#L428)  where developed to reduce the size of the covariance matrix removing the zero eigenvalues and then to restore it. These methods (now used in method `invert`) were previously used for the covariance decomposition. Should they be put back?

**Reply by EnricaBelfiore** [reply]
Are we considering the **UDU.T** decomposition in the LU decomposition? 

The method `to_positive` should rather be called `to_semipositive`, in my opinion, and it should be working. 
One could consider implementing the _real_ Cholesky decompositon to `CategoryCov.to_positive()._reduce_size()`, which is positive definite, and to use `_restore_size` to get back to the full matrix. Otherwise, we can use the square root Cholesky decomposition (**UDU.T**) to decompose the semi-positive approximation of the covariance matrix.

Note: the method `to_positive()` is here considered to be the one implemented, which returns a semi-positive definite matrix.

The definition of a proper measure for the magnitude of the error introduced by the matrix approximation would be needed to compare the two proposed methods.

**Reply by AitorBengoechea** [reply]
It will not work. If you reduce size, do cholesky decomposition and then restore size. If you do L*L^T with the restore size matrix, it will not return the original covariance matrix

**Comment by AitorBengoechea** [comment]
The problem is that there are rows full of zeros, so the determinant is automatically zero. 
The qr decomposition method automatically removed them, because it worked with eigenvalues and eigenvectors. But, cholesky uses the entire matrix.

**Reply by luca-fiorito-11** [reply]
Anyways, none of the considered methods actually use the cholesky decomposition

**Comment by luca-fiorito-11** [comment]
Can we agree that we need methods:

> to convert the covariance matrix into a semi-positive definite one
> to convert the semi-positive covariance e matrix into a positive covariance matrix
> to decompose the positive definite covariance matrix and extract L (for this step all decomposition methods should be equivalent, right?!)
> to restore the covariance size if it changed

Please, provide feedback

**Reply by AitorBengoechea** [reply]
With the qr decomposition method you obtain almost the same cov matrix. (test: 637X637 matrix)

**Reply by luca-fiorito-11** [reply]
What does "almost the same" mean?

**Reply by GrimFe** [reply]
In principle I agree with you @luca-fiorito-11. Still I think we should consider the possibility of decomposing semi-positive definite matrices with some of the methods mentioned above.
I think all the methods should converge to the same decomposition when a positive definite matrix is handled, indeed.
Since we are considering different approximations, I think help in evaluating the discrepancy they introduce and in comparing the approximations will be needed.

In fewer words: does decomposing a positive definite matrix improve the approximation or is this method a burden not really improving the code performance? Is semi-positive definite matrix decomposition good _enough_ or even better than positive definite one (where _enough_ depends on the discrepancy evaluation I mentioned above)?

**Reply by AitorBengoechea** [reply]
I though numerical fluctuations appear. Original matrix:
![Capture](https://user-images.githubusercontent.com/90201642/154662431-7b4999a9-997b-4639-89c5-716454ed736f.PNG)
 Calculated covariance matrix: 
![Capture(1)](https://user-images.githubusercontent.com/90201642/154662522-a2173704-92ec-4271-bb48-1ec90f444e64.PNG)

**Reply by AitorBengoechea** [reply]
Sorry, i have a mistake uploading the images. The below image is the original matrix

**Reply by EnricaBelfiore** [reply]
I checked with @GrimFe and we are not sure that for our purpose all the decomposition methods are equivalent. 

- LU decomposition for simmetric and positive definite matrix should converge to the Cholesky LL.T factorization, as both represent the original matrix (A from now on) as multiplication of lower and upper triangular matrices;
- QR decomposition will not converge to the Cholesky LL.T factorization because it represents A like an upper triangular matrix and a squared matrix Q, so if L = R.T, the product L*L.T will not return A.
  
So I need a clarification: what we need L to be? Just a lower triangular matrix or the one that satisfies the  Cholesky LL.T factorization?

**Reply by luca-fiorito-11** [reply]
It must satisfy `L^T L = C`. I believe this was the case for all mentioned decomposition, as they were implemented

**Reply by EnricaBelfiore** [reply]
Ok we should try also with the QR factorization but not for the covariance matrix. 
To be clear: we should perform this decomposition for a matrix that will be equal to  `eigenvector * sqrt(eigenvalues)` (to be consistent with Cholesky) of the covariance matrix. In this way the R.T obained will be equal to the L matrix that satisfies the requested constraints. If you have the previous code of sandy, can you please check if the QR factorization was performed in this way or in a similar way? Otherwise it cannot converge to the Cholesky factorization for me. Let me know if it is correct

**Reply by luca-fiorito-11** [reply]
The code is in the link above

**Reply by luca-fiorito-11** [reply]
Sorry, you can find the QR in line 1983 of https://github.com/luca-fiorito-11/sandy/blob/master/sandy/formats/utils.py

**Reply by AitorBengoechea** [reply]
Be careful with this method, it loses precision as the problem scales to a larger size. In my case, in 600 X 600 matrices I could already see numerical fluctuations.

**Reply by EnricaBelfiore** [reply]
Thank you for the link! It is consistent with what I wrote in my previous comment and it solves my doubts.
I had found the explanation in the following link: https://scicomp.stackexchange.com/questions/34645/how-can-i-get-cholesky-decomposition-from-eigenvalue-decomposition  
I hope it will be useful for your implementation @AitorBengoechea

**Comment by AitorBengoechea** [comment]
qr decomposition is not working for a real covariance matrix (Pu-241 compute with njoy for 241 groups with a covariance matrix 11809 X 11809). There is a lot of numerical fluctuation.

**Comment by luca-fiorito-11** [comment]
@AitorBengoechea , can you fix the implementation?  Only one method. If sparsity is too difficult to implement we neglect it for the moment.
This PR will follow the one on eigenvalues by @EnricaBelfiore .

To make this clear, the purpose of the `eig` method is ro return the eigenvalues and eigenvectors. In that sense, the existing implementation is ok as long as the parameter k is taken  into account (for instance as a kwarg). This implementation should in principle be independent from the decomposition. 

The decomposition could call the eig method if needed. The eig method could be adjusted if the decomposition cannot handle the existing implementation.  In that case the justification should be reported in the eig docstring.

**Comment by AitorBengoechea** [comment]
Thanks to @EnricaBelfiore  and @GrimFe for the help.

For the moment, with the methods available in sandy.cov and the qr decomposition (introducing a threshold and sparse calculations),  the results of the decomposition and reconstruction of the covariance matrix is equal to the original one (size 637 X 637)

**Reply by luca-fiorito-11** [reply]
But with what matrix?

**Reply by AitorBengoechea** [reply]
Pu-241 covariance matrix generated with njoy for 12 energy groups

**Reply by GrimFe** [reply]
Not having NJOY installed, I have been working with covariance matrices from endf6 files. Is this fine or is one forced to process covariance matrices in NJOY?

**Comment by luca-fiorito-11** [comment]
@EnricaBelfiore  and @GrimFe , is this issue fully associated to the eigenvalues problem? If we change the eig method, will this still be an issue. 

Can we clearly state what the problem is here?

**Reply by GrimFe** [reply]
In our opinion the two problems are different, even being related.

On the one hand, it is true that decomposition methods use `sandy.CategoryCov.eig`, which is expected to pose some decomposition issues as long as it is not fixes, and @EnricaBelfiore proposed now a version which should be stable, even if not based on sparse matrices.

On the other hand, the error we got was due to matrix singularity. This comes into play when performing LU decomposition, which is only possible for invertible matrices. Matrix invertibility is not only related to `sandy.CategoryCov.eig`, but rather to how nuclear data covariance matrices are produced and handled before decomposition. This means that before LU decomposing the covariance matrix one should approximate it in such a way that it is invertible.

In my opinion, the procedure to get L such that A=LL.T should look  like as follows:
```
def get_L(self):
        if positive_definite:
#             L = Cholesky - Which should be exactly what we are looking for and should hold as long as A is positive definite
            pass
        elif semipositive_definite:
#             L = UDU - Which should be Cholesky approximation for semipositive matrix decomposition
            pass
        else:
#             L = other - Here best decomposition approximation should be implemented*
            pass
        return L
```
\* Note: one should here implement one of the following options, I think (more options could of course be considered):

1. 
- transform the covariance matrix into semi-positive definite one with a proper method (this should now be handled by `to_positive`, but I am not completely sure of its correct behavior since I got some strange results - I will check as soon as possible more in the details as I might very well have done something wrong)
- Decompose the now-semi-positive definite matrix with UDU method used in the previous if section (`elif semipositive_definite`).
2.
- transform the covariance matrix into semi-positive definite one with a proper method (this should now be handled by `to_positive`, but I am not completely sure of its correct behavior since I got some strange results - I will check as soon as possible more in the details as I might very well have done something wrong)
- transform the now-semi-positive definite matrix into a positive definite with a proper method removing rows and columns corresponding to negative eigenvalues, as suggested by @luca-fiorito-11 (please note that negative eigenvalues should be well handled by this two steps procedure)
- Decompose the now-positive definite matrix with Cholesky method used in the first if section (`if positive_definite`).
3.
- Implement completely new matrix factorization, such as QR, which should be done according to what present in previous SANDY version and reported by @EnricaBelfiore.

I am working on a notebook to get insights on which approximation gives the best representation of the eigenvalues of the covariance matrix. I will share it with you, @AitorBengoechea, as soon as I get relevant results as discussed this morning on Teams.
Please also note that an option forcing the method to use QR (which could be the most precise even with positive matrices, according to my first results, but which is also the most computationally expensive) could be considered.

We hope this clarifies our point of view on the problem and how we think it might be tackled.

@luca-fiorito-11 is this a good approach in your opinion? Please give a feedback so that we can focus our work and make it as less scattered as possible.

**Reply by luca-fiorito-11** [reply]
Options 1 should be the only selected option.
We don't want to maintain two methods!
If the UDU decomposition works both for positive-definite and for semi-positive definite matrices, then it should be the only method maintained.

This will require that `to_positive` is correct. @GrimFe , since you are looking into it, can I expect from you a PR with the necessary changes (if needed) and tests?

For my info, these options will potentially make `reduce_size` and `restore_size` obsolete, right?

**Reply by GrimFe** [reply]
If we decide that the way to go is option 1, then there is no need of `reduce_size` nor of `restore_size` for this method.
I will share updates on `to_positive` as soon as I get them.

**Reply by AitorBengoechea** [reply]
For the moment, no matter with method we use, qr decomposition or scipy.linalg.ldl, both have the same problem: when the size of the covariance matrix increase, the method not produce the same results. 
For small matrix (Pu-241 covariance matrix generated with njoy for 12 energy groups, shape=(637, 637)) is better to use scipy.linalg.ldl, because the maintenance is done by scipy. For big matrixs (Pu-241 covariance matrix generated with njoy for 241 energy groups, shape=(11809, 11809) ), the method qr produces better results in all the matrix.

**Reply by AitorBengoechea** [reply]
I mean, comparing the original covariance matrix and the reconstructed one

**Reply by GrimFe** [reply]
If QR needs to be used for bigger matrices (i.e. more computationally expensive cases), why aren't we using QR for smaller matrices?

**Reply by AitorBengoechea** [reply]
Because i only put the conclusions, for the moment, both methods have problems when the size is increased

**Reply by luca-fiorito-11** [reply]
@AitorBengoechea , considering size problems and everything...what is in your opinion THE best method to implement? (only one)

**Reply by GrimFe** [reply]
> This will require that to_positive is correct. @GrimFe , since you are looking into it, can I expect from you a PR with the necessary changes (if needed) and tests?


Concerning the issues on `to_positive()`:
What I get is `TypeError: Covariance matrix must be symmetric` when running
```
tape = sandy.get_endf6_file("jeff_33", "xs", 942390) # Pu239
Cov = sandy.XsCov.from_endf6(tape)
cov = sandy.CategoryCov(Cov)
P = cov.to_positive()
```
This seems to be related to the tolerance limit considered for the matrix symmetry check (`(data.values.round(14) == data.values.T.round(14)).all()` in `@data.setter`).
Rounding at 12 solves the problem for Pu239, but the same issue might raise for other nuclides.
Before thinking of changing such rounding value, I was wondering where it come from. Is it the precision of data in ENDF6 file? In case not, why are we not sticking to that precision? In case yes, maybe `to_positive()` should be a function returning a `pd.DataFrame()` object which can not be used as builder for `CategoryCov()` object, being it not symmetric.
When considering pros and cons of different covariance matrix decomposition methods, I think one should also consider that implementation and maintenance of `to_positive()` (and of `to_semipositive()`, if ever implemented/used to replace the current `to_positive()`) methods is not needed if **QR** decomposition is implemented, as it should work for all matrices.


One more question: why are we using sparse matrices? I understand it enhances the possibilities of SANDY, allowing one to handle much bigger matrices, but I do not know what is the typical size of matrices handled by `CategoryCov()`. I mean, 
```
tape = sandy.get_endf6_file("jeff_33", "xs", 942390) # Pu239
Cov = sandy.XsCov.from_endf6(tape)
cov = sandy.CategoryCov(Cov)
cov.data.shape
```
gives (295, 295).
Do you expect much bigger matrices to be handled by `CategoryCov()`?

**Reply by AitorBengoechea** [reply]
A real covariance matrix, for 241 groups of energy generated with njoy has a shape (11809 X 11809). Q matrix or B matrix taken from the decay data have 4000 X 4000 ... and inside category Cov you have the GLS method for the covariance. In this case, only taking fission yields, you are working with a diagonal matrix of 1000 X 1000. 
The numpy and pandas algorithm works well for matrices smaller than 4000 X 4000. Also you dont have to take care of sparse matrices, sparse_tables functions are already implemented for doing multiplication and inversion of very big matrices.

**Reply by GrimFe** [reply]
Ok, thank you, I see your point.
My question was related to the fact that not all decompositions are implemented in `scipy.linalg.sparse` module, if I am not mistaken. And I am not sure switching from sparse matrices to dense ones helps code performances and readability. For this reason, this should also be considered when it comes to the identification of the _best_ decomposition technique.

**Reply by luca-fiorito-11** [reply]
Let's not forget our goal. We need __one__ method to decompose the covariance matrix. This method was the QR decomposition for dense matrices that was implemented and which is still available on the master branch. Effiency is problem dependent and can be considered later. Sparsity is preferable but not needed at the moment. Other methods such as `to_positive` etc. mean more maintenance. They will be kept only if needed. If the QR decomposition is still the preferable one, then we'll revert to what we had.

**Reply by luca-fiorito-11** [reply]
About the symmetry check. This check is welcome but it cannot become a burden. There is no reason for checking 14 digits. We could check less or none, or we could weaken the error to a warning. The point is that in no way the check should become a problem. Maybe we should just remove since in the development of SANDY we start from the principle that "the users of the API know what they are doing"....then they will not load a no symmetrical matrix to CategoryCov

**Reply by GrimFe** [reply]
@AitorBengoechea FYI: this (https://github.com/yig/PySPQR) might be relevant for sparse matrices-based optimization in case we revert to **QR**.

**Comment by AitorBengoechea** [comment]
@EnricaBelfiore , @luca-fiorito-11 and @GrimFe :

Decomposition problem solved. We are going to select the qr decomposition. My results are the following:

Problem size and description: Pu-241 xs computed with njoy for 241 energy groups with a covariance matrix 11809 X 11809

Eigenvalues and eigenvectors obtained in 8-10 minutes

A = V.dot(E), then the transpose and finally the qr decomposition: between 20s to 2 min

Reconstruction of the covariance matrix: between 3 min to 4 min

Results: comparison of the relative difference of the original covariance matrix and the reconstructed one:
- The first values (more accurate because they perform less operations due to the triangular matrix structure): highest relative difference of 4.31880e-06 %
- The last values ( less accurate because there are more operation in the calculations): highest relative difference of 1.19921e-02 %

**Reply by GrimFe** [reply]
@AitorBengoechea 
I am not quite sure I understand what you mean by
> A = V.dot(E), then the transpose and finally the qr decomposition

Is A the matrix to **QR** decompose in order to get L == R.T, where R comes from **QR** decomposition and L is such that Cov == LL.T?

**Comment by luca-fiorito-11** [comment]
- the matrix size being 11809, and considering 241 energy groups, means  11809/241=49 MT reactions. This is a large number of reactions,  is it true? 2nd note, even if we consider that pu241 has covariance data for 49 MTs, most of the associated reactions should have a threshold, meaning that many variance terms should be zero. This could in principle be largely simplified. To be checked if these zero terms affect the efficiency of the eigenvalue calculation.
- 10 minutes for the eig calculation is a lot. I don't recall having such problems with the master branch 
- to set up a reference, can we assume ~20000 as the largest matrix size that we need to handle at the moment?

**Reply by AitorBengoechea** [reply]
If i do using reduce size and restore size, the precision doesnt change, but the times are much lower than with all the matrix:
Matrix size reduction: from 11809 to 6283

Eigenvalues and eigenvectors obtained in 1-2 minutes

A = V.dot(E), then the transpose and finally the qr decomposition: approx 30s

Reconstruction of the covariance matrix: between 2min

**Comment by luca-fiorito-11** [comment]
FYI, have a look at the [Wiki](https://en.m.wikipedia.org/wiki/Cholesky_decomposition)  page about cholesky and its proof by QR decomposition

**Comment by luca-fiorito-11** [comment]
@AitorBengoechea , `CategoryCov.eig` method was updated with PR #139.
You can push your changes for the covariance decomposition.
